File: src/station.java
import java.io.*;
import java.net.*;
import java.nio.*;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.text.SimpleDateFormat;
import java.util.*;
import java.nio.channels.*;
import java.nio.charset.StandardCharsets;

public class station{

    static String station_name;
    static int tcp_port;
    static int udp_port;
    //Stores adjacent port Numbers
    static int[] neighbor_port;
    static double[] location = new double[2];
    

    static String[][] timetable;
    
    static LinkedList<Integer> neighbor = new LinkedList<Integer>();
    //Store the final time to the station in order to find the fastest path
    static LinkedList<String> time_compare = new LinkedList<String>();
    //Store the output ready to be sent to the browser to find the fastest path
    static Map<String,String> save_string = new HashMap<String,String>();
    //The number of buses to the station
    static LinkedList<String> bus = new LinkedList<String>();
    //Store the output ready to be sent，To know which socket to send
    static Map<String,String> result = new HashMap<String,String>();
    //the dictionary which key is socket, value is destination,To know which socket to send
    static Map<SocketChannel,String> s = new HashMap<SocketChannel,String>();
    //Stores the number of paths from any starting point to the station to find the fastest path
    static Map<String,Integer> route_num = new HashMap<String,Integer>();


    //read timetable
    public static void read_timetable(String filename) throws Exception {

        File directory = new File("");
        String courseFile = directory.getCanonicalPath();
        String path = courseFile + "/" + "tt-"+ filename;

        InputStream is = new FileInputStream(path);
        BufferedReader br = new BufferedReader(new InputStreamReader(is));

        FileReader fileReader = new FileReader(path);
        LineNumberReader lineNumberReader = new LineNumberReader(fileReader);
        lineNumberReader.skip(Long.MAX_VALUE);
        long lines = lineNumberReader.getLineNumber() + 1;
        fileReader.close();
        lineNumberReader.close();

        timetable = new String[(int) lines - 2][5];

        String line;
        line = br.readLine();
        String[] arr = line.split(",");
        location[0] = Double.valueOf(arr[1]);
        location[1] = Double.valueOf(arr[2]);
        int j = 0;

        while ((line = br.readLine()) != null) {
            arr = line.split(",");
            timetable[j] = arr;
            
            j++;
        }

        br.close();

    }

    //Gets the last time the schedule was modified
    public static String get_modifytime(String filename) throws Exception{
        File directory = new File("");
        String courseFile = directory.getCanonicalPath();
        String path = courseFile + "/" + "tt-"+ filename;
        File file = new File(path);
        Long lastModified = file.lastModified();
        Date date = new Date(lastModified);
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String mo_time = formatter.format(date);
        return mo_time;
    }

    //check the timetable changed or not
    public static boolean is_timetableChange(String stationname,String now_time) throws Exception{
        String changed_time = get_modifytime(stationname);
        Integer bool = changed_time.compareTo(now_time);
        if(bool>0){
            return true;
        }
        else{
            return false;
        }

    }

    public static void main(String[] args) throws Exception{
        //Processing parameters
        station_name = args[0];
        tcp_port = Integer.valueOf(args[1]);
        udp_port = Integer.valueOf(args[2]);
        int m=0;
        neighbor_port = new int [args.length-3];
        
        for(int i=3;i<args.length;i++){
            neighbor_port[m]=Integer.valueOf(args[i]);
            neighbor.add(neighbor_port[m]);
            m++;
        }
        read_timetable(station_name);
        String modify_time = get_modifytime(station_name);

        //initialises its TCP and UDP ports
        //Create a service
        ServerSocketChannel server = ServerSocketChannel.open();
        //Specified as non-blocking, default to blocking
        server.configureBlocking(false);
        //create channel
        DatagramChannel channel = DatagramChannel.open();
        channel.configureBlocking(false);
        DatagramSocket socket = channel.socket();

        //bind port
        server.socket().bind(new InetSocketAddress(tcp_port));
        //bind ip and port
        InetSocketAddress address = new InetSocketAddress(udp_port);
        socket.bind(address);

        //create selector
        Selector selector = Selector.open();
        Selector selector1 = Selector.open();

        //Register the connection ready event
        server.register(selector, SelectionKey.OP_ACCEPT);
        channel.register(selector1, SelectionKey.OP_READ);


        while(true){
            neighbor = new LinkedList<Integer>();
            for(int j=0;j<neighbor_port.length;j++){
                neighbor.add(neighbor_port[j]);
            }

            //Found the event
            if (selector1.select(1000) < 1 && selector.select(1000)<1) {
                continue;
            }
            //Put the selected into the set and iterate
            Set<SelectionKey> keys = selector.selectedKeys();
            Set<SelectionKey> keys1 = selector1.selectedKeys();
            Iterator<SelectionKey> it = keys.iterator();
            Iterator<SelectionKey> it1 = keys1.iterator();

            while (it.hasNext()) {
                SelectionKey key = it.next();
                it.remove();
                if (key.isAcceptable()) {  //Receive events
                    //Gets the key associated with the ServerSocketChannel
                    ServerSocketChannel ssc = (ServerSocketChannel) key.channel();
                    //Gets the SocketChannel for the key connection
                    SocketChannel sc = ssc.accept();
                    sc.configureBlocking(false);
                    //The request is received, the connection is registered to the selector, and the listener type is set to read
                    sc.register(selector,SelectionKey.OP_READ);
                }
                else if(key.isReadable()){
                    //check modified or not
                    if(is_timetableChange(station_name,modify_time)){
                        read_timetable(station_name);
                        modify_time = get_modifytime(station_name);
                    }

                    //get destination
                    SocketChannel ch = (SocketChannel)key.channel();
                    ByteBuffer dest = ByteBuffer.allocate(1024);
                    int read = ch.read(dest);
                    String data = new String(dest.array(),0,read);
                    String[] arr = data.split("\n");
                    String line = arr[0];
                    String[] array = line.split(" ");
                    String destination = array[1].substring(5);

                    //avoid favicon
                    String ico = "con.ico";
                    if(destination.equals(ico)){
                        ch.close();
                        destination = "";
                        continue;
                    }
                    
                    result.put(destination, null);
                    s.put(ch,destination);
                        
                    //create data_send,This data means find route
                    //The format is: 1; destination; route
                    String string_send = String.valueOf(1)+ ";" + destination + ";" + station_name + "," + String.valueOf(udp_port);
                    byte[] bys = string_send.getBytes();
                        
                    for(int i=0;i<neighbor_port.length;i++){
                        ByteBuffer buf=ByteBuffer.allocate(1024);
                        buf.clear();
                        buf.put(bys);
                        buf.flip();
                        channel.send(buf, new InetSocketAddress("127.0.0.1",neighbor_port[i]));
                    }
                    //skt remove from listen append write
                    ch.register(selector, SelectionKey.OP_WRITE);
                    
                    


                }
                else if(key.isWritable()){
                    SocketChannel ch =(SocketChannel)key.channel();
                    String destination = s.get(ch);
                    if(result.get(destination)==null){
                        //not prepare for write
                        continue;
                    }
                    if(result.get(destination)!=null){
                        //http response
                        String output = result.get(destination);

                        ch.write(ByteBuffer.wrap("HTTP/1.1 200 OK\r\n".getBytes()));
                        ch.write(ByteBuffer.wrap("Content-Type:text/plain\r\n".getBytes()));
                        ch.write(ByteBuffer.wrap("\r\n".getBytes()));
                        ch.write(ByteBuffer.wrap(output.getBytes()));
                        ch.close();
                        
                    }
                }
            }

            while(it1.hasNext()){
                //udp event
                SelectionKey key1 = it1.next();
                it1.remove();
                if(key1.isReadable()){
                    DatagramChannel datagramChannel =(DatagramChannel)key1.channel();
                    //decode data
                    ByteBuffer readBuffer = ByteBuffer.allocate(1024);
                    readBuffer.clear();
                    datagramChannel.receive(readBuffer);
                    readBuffer.flip();
                    CharBuffer charBuffer = StandardCharsets.UTF_8.decode(readBuffer); 
                    String data_string = charBuffer.toString(); 
                    //Split the received information
                    String[] arr = data_string.split(";");
                    
                    
                    System.out.println("receive:"+data_string);
                        //Check the prefix
                        if(Integer.parseInt(arr[0]) == 1){
                            System.out.println("find route");

                            int a = arr.length-1;

                            if(arr[1].equals(station_name)){//Whether to find the destination or not
                                if(route_num.containsKey(arr[2].split(",")[0])){//record the num of routes from this start point to the destination
                                    if(route_num.get(arr[2].split(",")[0])==null){
                                        break;
                                    }
                                    else{
                                        int route = route_num.get(arr[2].split(",")[0]);
                                        route_num.replace(arr[2].split(",")[0],route++);
                                    }
                                    
                                }
                                else{
                                    route_num.put(arr[2].split(",")[0],1);
                                }
                                
                                //create data_send,This data means sending the path back to the starting point
                                //The format is: 2; The starting point; The path; The port number of the current station
                                String data_send = String.valueOf(2) + ";" + arr[2].split(",")[0] + ";";
                                for(int j=2;j<=a;j++){
                                    data_send = data_send + arr[j] + ";";
                                }
                                data_send = data_send + station_name + "," + udp_port + ";" + String.valueOf(arr.length);
                                int port = Integer.parseInt(arr[a].split(",")[1]);

                                byte[] by = data_send.getBytes(StandardCharsets.UTF_8); 
                                ByteBuffer buf=ByteBuffer.allocate(1024);
		                        buf.clear();
		                        buf.put(by);
		                        buf.flip();
                                channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                System.out.println("send to:"+ port);
                                System.out.println("send message is:"+data_send);
                                

                            }
                            if(!arr[1].equals(station_name)){//not find the destination
                                for(int j=2;j<=a;j++){//Check which stations exist in the path,Are these stations adjacent to the current station?
                                    if(neighbor.contains(Integer.parseInt(arr[j].split(",")[1]))){
                                        Object o = Integer.parseInt(arr[j].split(",")[1]);
                                        neighbor.remove(o);
                                    }
                                }
                                //Add information about the current station
                                String data_send = new String();
                                for(int j=0;j<=a;j++){
                                    data_send=data_send + arr[j] + ";";
                                }
                                data_send = data_send + station_name + "," + udp_port;

                                byte[] by = data_send.getBytes(StandardCharsets.UTF_8); 
          
                                int size = neighbor.size();
                                for(int j=0;j<size;j++){
                                    ByteBuffer buf=ByteBuffer.allocate(1024);
                                    buf.clear();
                                    buf.put(by);
                                    buf.flip();
                                    System.out.println("buf:"+ buf);
                                    System.out.println("j="+j);
                                    int port = (int) neighbor.remove();
                                    channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                    System.out.println("send to:"+ port);
                                    System.out.println("send message is:"+data_send);
                                    
                                }
                                
                                
                            }
                        }
                        if(Integer.parseInt(arr[0]) == 2){//Return the route
                            System.out.println("Return the route");
                            
                            int a = arr.length-2;
                            if(arr[1].equals(station_name)){
                                
                                int b = Integer.parseInt(arr[arr.length-1]);
                                //find time
                                SimpleDateFormat sdf = new SimpleDateFormat();//Formatting time  
                                sdf.applyPattern("HH:mm"); 
                                Date now = new Date();// get current time
                                String now_string = sdf.format(now);
                                
                                String time = new String();
                                String time_inital = new String();
                                String bus = new String();

                                //find the next bus
                                for(int j=0;j<timetable.length;j++){
                                    
                                    Integer bool = timetable[j][0].compareTo(now_string);

                                    if(bool >= 0){
                                        if(timetable[j][4].equals(arr[b].split(",")[0])){
                                    
                                            time = timetable[j][3];
                                            time_inital = timetable[j][0];
                                            bus = timetable[j][1];
                                            break;
                                        }
                                        
                                    }
                                    
                                }

                                //create data_send,This data means find time
                                //The format is: 3; time; time_inital; route; bus; The port number of the current station
                                String data_send = String.valueOf(3) + ";" + time + ";" + time_inital + ";";
                                for(int j=2;j<=a;j++){
                                    data_send = data_send + arr[j] + ";";
                                }
                                data_send = data_send + bus + ";" + String.valueOf(3);

                                int port = Integer.parseInt(arr[3].split(",")[1]);
                                System.out.println("send to:"+ port);
                                System.out.println("send message is:"+data_send);
                                byte[] by = data_send.getBytes(); 
                                ByteBuffer buf=ByteBuffer.allocate(1024);
		                        buf.clear();
		                        buf.put(by);
		                        buf.flip();
                                channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                

                            }
                            if(!arr[1].equals(station_name)){//not find the start
                                System.out.println("not find the destnation");
                                String data_send = new String();
                                for(int j=0;j<=a;j++){
                                    data_send=data_send + arr[j] + ";";
                                }
                                int b = Integer.valueOf(arr[arr.length-1]);
                                data_send = data_send + String.valueOf(b-1);

                                int port = Integer.parseInt(arr[b-2].split(",")[1]);
                                System.out.println("send to:"+ port);
                                System.out.println("send message is:"+data_send);

                                byte[] by = data_send.getBytes(StandardCharsets.UTF_8); 
                                ByteBuffer buf=ByteBuffer.allocate(1024);
		                        buf.clear();
		                        buf.put(by);
		                        buf.flip();
                                channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                
                                
                            }
                        }
                        if(Integer.parseInt(arr[0]) == 3){
                            System.out.println("find time");
                            
                            if(arr[arr.length-3].split(",")[0].equals(station_name)){//find the destnation
                                System.out.println("find the destnation");
                                String bus_num = arr[arr.length-2];
                                String time = arr[1];
                                //create data_send,This data means return time
                                //The format is: 4; start station; time; time_inital; route; The port number of the current station
                                String data_send = String.valueOf(4) + ";" + arr[3].split(",")[0] + ";" + arr[2] + ";" + time + ";";
                                for(int j = 3;j<=arr.length-3;j++){
                                    data_send=data_send+arr[j]+";";
                                }
                                data_send = data_send + String.valueOf(arr.length-2);
                                String start = arr[3].split(",")[0];
                                
                                //The data has been sent
                                if(route_num.get(start)==null){
                                    break;
                                }
                                //begin to compare the time
                                if(bus.size()==route_num.get(start)-1){
                                    while(!time_compare.isEmpty()){
                                        String remove = time_compare.remove();
                                        Integer bool = time.compareTo(remove);
                                        if(bool>0){
                                            data_send = save_string.get(remove);
                                            time = remove;
                                        }
                                    }
                                    String[] arr2 = data_send.split(";");
                                    int b1 = Integer.parseInt(arr2[arr2.length-1])-1;
                                    
                                    int port = Integer.parseInt(arr2[b1].split(",")[1]);
                                    
                                    
                                    bus.clear();
                                    time_compare.clear();
                                    save_string.clear();
                                    route_num.replace(start,null);
                                    byte[] by = data_send.getBytes(StandardCharsets.UTF_8); 
                                    ByteBuffer buf=ByteBuffer.allocate(1024);
                                    buf.clear();
                                    buf.put(by);
                                    buf.flip();
                                    channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                    System.out.println("send to:"+ port);
                                    System.out.println("send message is:"+data_send);
                                }
                                else{
                                    if(bus.contains(bus_num)){
                                        bus.add(bus_num);
                                        time_compare.add(time);
                                        save_string.put(time,data_send);
                                    }
                                }
                            }
                            if(!arr[arr.length-3].split(",")[0].equals(station_name)){//not find the destnation
                                System.out.println("not find the destnation");
                                String data_send = String.valueOf(3) + ";";
                                String now = arr[1];
                                String bus = new String();
                                int b = Integer.parseInt(arr[arr.length-1]);
                                //find next bus
                                for(int j=0;j<timetable.length;j++){
                                    Integer bool = timetable[j][0].compareTo(now);
                                    if(bool >= 0){
                                        if(timetable[j][4].equals(arr[b+2].split(",")[0])){
                                            now = timetable[j][3];
                                            bus = timetable[j][1];
                                            break;
                                        }
                                        
                                    }
                                }
                                data_send = data_send + now + ";" + arr[2] + ";";
                                for(int j=3;j<=arr.length-3;j++){
                                    data_send = data_send + arr[j] +";";
                                }
                                
                                data_send = data_send + bus+ ";" + String.valueOf(b+1);
                                int port = Integer.parseInt(arr[b+2].split(",")[1]);
                                System.out.println("send to:"+ port);
                                System.out.println("send message is:"+data_send);

                                byte[] by = data_send.getBytes(); 
                                ByteBuffer buf=ByteBuffer.allocate(1024);
		                        buf.clear();
		                        buf.put(by);
		                        buf.flip();
                                channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                                
                            }
                        }
                        if(Integer.parseInt(arr[0]) == 4){
                            System.out.println("return time");
                            if(arr[1].equals(station_name)){
                                //start preparing for the HTTP response
                                String time_arrive = arr[3];
                                String bus = new String();
                                String stop = new String();
                                String time = arr[2];
                                String output = new String();
                                for(int j=0;j<timetable.length;j++){
                                    
                                    if(timetable[j][0].equals(time)){
                                        bus = timetable[j][1];
                                        stop = timetable[j][2];
                                        break;
                                    }
                                }
                                if(time_arrive.equals("")||bus.equals("")||stop.equals("")){
                                    //can not find the bus today
                                    output = "a valid route does not exist on the current day.";
                                }
                                else{
                                    output = "At " + time + ", catch " + bus + ", from " + stop +". You will arrive at your final destination at "+ time_arrive +".";
                                }
                                result.put(arr[arr.length-2].split(",")[0], output);

                                

                            }
                            if(!arr[1].equals(station_name)){
                                //not find the start
                                String data_send = new String();
                                for(int j=0;j<=arr.length-2;j++){
                                    data_send=data_send + arr[j] + ";";
                                }
                                int b = Integer.valueOf(arr[arr.length-1]);
                                data_send = data_send + String.valueOf(b-1);                                

                                int port = Integer.parseInt(arr[b-2].split(",")[1]);
                                System.out.println("send to:"+ port);
                                System.out.println("send message is:"+data_send);
                                byte[] by = data_send.getBytes(StandardCharsets.UTF_8); 
                                ByteBuffer buf=ByteBuffer.allocate(1024);
		                        buf.clear();
		                        buf.put(by);
		                        buf.flip();
                                channel.send(buf, new InetSocketAddress("127.0.0.1",port));
                            }
                        


                    }
                }
            }
            
            

        }
    }
}

File: src/station.py
import sys
import socket
import select
import time
import re
import os


def readtimetable(station_name):
    timetable = []
    filename = 'tt-' + station_name
    with open(filename) as file:
        for line in file.readlines():
            line = line.strip('\n')
            info = line.split(',')
            timetable.append(info)
    file.close()
    return timetable

def is_timetableChange(stationname,now_time):
    changed_time = time.ctime(os.stat('tt-'+stationname).st_mtime)
    if now_time < changed_time:
        return True
    else:
        return False

def main():
    #Processing parameters
    if len(sys.argv)<5:
        sys.ecit()
    
    station_name = sys.argv[1]
    tcp_port = int(sys.argv[2])
    udp_port = int(sys.argv[3])

    #Stores adjacent port Numbers
    neighbour_port = []
    #Stores the number of paths from any starting point to the station to find the fastest path
    route_num = {}
    #Store the output ready to be sent to the browser to find the fastest path
    save_string = {}
    #The number of buses to the station
    Bus = []
    #Store the final time to the station in order to find the fastest path
    time_compare = []
    #Store the output ready to be sent，To know which socket to send
    result = {}
    #the dictionary which key is socket, value is destination,To know which socket to send
    s = {}
    localhost = '127.0.0.1'
    
    for i in range(4,len(sys.argv)):
        neibour = int(sys.argv[i])
        neighbour_port.append(neibour)
    
    timetable = readtimetable(station_name)
    #Gets the last time the schedule was modified
    modify_time = time.ctime(os.stat('tt-'+station_name).st_mtime)

    #initialises its TCP and UDP ports
    tcpsocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    tcpsocket.bind((localhost,tcp_port))
    tcpsocket.listen()
    tcpsocket.setblocking(False)
    udpsocket = socket.socket(socket.AF_INET,socket.SOCK_DGRAM)
    udpsocket.bind((localhost,udp_port))

    #create listenlist and writelist
    listenlist = [tcpsocket,udpsocket,]
    writelist = []

    while True:
        neighbour = []
        for i in neighbour_port:
            ne = i;
            neighbour.append(ne)

        #use select
        read,write,error = select.select(listenlist,writelist,[])

        for skt in read:
            if skt == tcpsocket:
                #tcp accept
                conn, addr = tcpsocket.accept()
                conn.setblocking(False)
                listenlist.append(conn)
            elif skt == udpsocket:
                #udp read
                data, address = udpsocket.recvfrom(1024)
                #decode data
                data_string = data.decode('utf-8')
                print('receive:'+ data_string)
                #Split the received information
                arr = data_string.split(';')

                #Check the prefix
                if int(arr[0]) == 1:
                    print('find route')

                    #Whether to find the destination or not
                    if arr[1]==station_name:
                        print('find the destination')
                        #get the start staion of this route
                        start = arr[2].split(',')[0]

                        #record the num of routes from this start point to the destination
                        if start in route_num:
                            if route_num[start] == None:
                                break
                            else:
                                route = route_num[start]
                                route_num[start] = route+1
                        else:
                            route_num[start] = 1
                        
                        #create data_send,This data means sending the path back to the starting point
                        data_send=''
                        #The format is: 2; The starting point; The path; The port number of the current station
                        data_send = str(2) + ';' + arr[2].split(',')[0] + ';'
                        for j in range(2,len(arr)):
                            data_send = data_send + arr[j] + ';'

                        data_send = data_send + station_name + ',' + str(udp_port) + ';' + str(len(arr))
                        
                        port = int(arr[len(arr)-1].split(',')[1])
                        print('send to:'+ str(port))
                        print('send message is：'+ data_send)
                        skt.sendto(data_send.encode('utf-8'),(localhost,port))
                    elif arr[1]!=station_name:
                        print('not find the destination')

                        #Check which stations exist in the path,Are these stations adjacent to the current station?
                        for j in range(2,len(arr)):
                            n = arr[j].split(',')[1]
                            int_n = int(n)
                            for i in neighbour:
                                if int_n == i:
                                    neighbour.remove(int_n)
                        
                        #Add information about the current station
                        data_send=''
                        for j in range(0,len(arr)):
                            data_send = data_send + arr[j] + ';'
                        data_send = data_send + station_name + ',' + str(udp_port)

                        
                        for j in range(0,len(neighbour)):
                            port = neighbour.pop()
                            print('send to:'+ str(port))
                            print('send message is：'+ data_send)
                            
                            skt.sendto(data_send.encode('utf-8'),(localhost,port))
                elif int(arr[0])==2:
                    print('Return the route')
                    if arr[1]==station_name:
                        print('find the start')
                        b = int(arr[len(arr)-1])
                        #get now time
                        now = time.strftime("%H:%M")
                        time_inital = ''
                        bus = ''
                        Time = ''
                        #find the next bus
                        for j in range(1,len(timetable)):
                            if timetable[j][0]>now and timetable[j][4] == arr[b].split(',')[0] :
                                Time = timetable[j][3]
                                time_inital = timetable[j][0]
                                bus = timetable[j][1]
                                break
                        #create data_send,This data means find time
                        data_send=''
                        #The format is: 3; time; time_inital; route; bus; The port number of the current station 
                        data_send = str(3) + ';' + Time + ';' + time_inital + ';'
                        for j in range(2,len(arr)-1):
                            data_send =data_send +arr[j] + ';'
                        
                        data_send = data_send +bus + ';' + str(3)
                        port = int(arr[3].split(',')[1])
                        print('send to:'+ str(port))
                        print('send message is：'+ data_send)
                        skt.sendto(data_send.encode('utf-8'),(localhost,port))
                    elif arr[1]!=station_name:
                        print('not find the start')
                        data_send=''
                        for j in range(0,len(arr)-1):
                            data_send = data_send + arr[j] + ';'
                        b = int(arr[len(arr)-1])
                        data_send = data_send + str(b-1)
                        port = int(arr[b-2].split(',')[1])
                        print('send to:'+ str(port))
                        print('send message is：'+ data_send)
                        skt.sendto(data_send.encode('utf-8'),(localhost,port))
                elif int(arr[0])==3:
                    print('find time')
                    
                    if arr[len(arr)-3].split(',')[0]==station_name:
                        print('find the destnation')
                        bus_num = arr[len(arr)-2]
                        Time = arr[1]
                        #create data_send,This data means return time
                        data_send=''
                        #The format is: 4; start station; time; time_inital; route; The port number of the current station
                        data_send = str(4) + ';' + arr[3].split(',')[0] + ';' + arr[2] + ';' +Time + ';'
                        for j in range(3,len(arr)-2):
                            data_send = data_send + arr[j] + ';'
                        data_send = data_send + str(len(arr)-2)
                        start = arr[3].split(',')[0]
                        
                        #The data has been sent
                        if route_num[start]==None:
                            break
                        #begin to compare the time
                        if len(Bus) == route_num[start]-1:
                            while len(time_compare)!=0:
                                remove = time_compare.pop()
                                if Time>remove :
                                    data_send = save_string[remove]
                                    Time = remove
                            arr1 = data_send.split(';')
                            b1 = int(arr1[len(arr1)-1])-1
                            port = int(arr1[b1].split(',')[1])
                            Bus.clear()
                            time_compare.clear()
                            save_string.clear()
                            route_num[start]=None
                            print('send to:'+str(port))
                            skt.sendto(data_send.encode('utf-8'),(localhost,port))
                        else:
                            if bus_num not in Bus:
                                Bus.append(bus_num)
                                time_compare.append(Time)
                                save_string[Time] = data_send
                    elif arr[len(arr)-3].split(',')[0]!=station_name:
                        print('not find the destnation')
                        data_send=''
                        data_send = str(3)+';'
                        now = arr[1]
                        BUS = ''
                        b = int(arr[len(arr)-1])
                        #find next bus
                        for j in range(1,len(timetable)):
                            if timetable[j][0]>now:
                                print('j='+str(j))
                                if timetable[j][4]==arr[b+2].split(',')[0]:
                                    now = timetable[j][3]
                                    BUS = timetable[j][1]
                                    break
                        data_send = data_send + now + ';' + arr[2] + ';'
                        for j in range(3,len(arr)-2):
                            data_send = data_send + arr[j] + ';'
                        
                        data_send = data_send +BUS+ ';' + str(b+1)
                        port = int(arr[b+2].split(',')[1])
                        print('send to:'+ str(port))
                        print('send message is：'+ data_send)
                        skt.sendto(data_send.encode('utf-8'),(localhost,port))
                elif int(arr[0])==4:
                    print('return time')
                    if arr[1]==station_name:
                        print('find the start')
                        #start preparing for the HTTP response
                        time_arrive = arr[3]
                        use_bus = ''
                        stop = ''
                        use_time = arr[2]
                        for j in range(0,len(timetable)):
                            if timetable[j][0]==use_time:
                                use_bus = timetable[j][1]
                                stop = timetable[j][2]
                                break
                        #can not find the bus today
                        if use_time=='' or time_arrive=='' or use_bus=='' or stop=='' :
                            output = 'a valid route does not exist on the current day.'
                        else:
                            output = 'At' + use_time + ', catch ' + use_bus + ', from ' + stop + '. You will arrive at your final destination at ' + time_arrive + '.'
                        result[arr[len(arr)-2].split(',')[0]] = output
                    elif arr[1]!=station_name:
                        print('not find the start')
                        data_send = ''
                        for j in range(0,len(arr)-1):
                            data_send = data_send + arr[j] + ';'
                        b = int(arr[len(arr)-1])
                        data_send = data_send + str(b-1)
                        port = int(arr[b-2].split(',')[1])
                        print('send to:'+ str(port))
                        print('send message is：'+ data_send)
                        skt.sendto(data_send.encode('utf-8'),(localhost,port))    
            else:
                #check timetable changed or not
                if is_timetableChange(station_name,modify_time):
                    readtimetable(station_name)
                    modify_time = time.ctime(os.stat('tt-'+station_name).st_mtime)
                #get the destination
                destination = ''
                data = skt.recv(1024).decode('utf-8')
                request_header_lines = data.splitlines()
                http_header_data = request_header_lines[0]
                data_list = re.split(r"/|\?\s|=| ",http_header_data)
                for i in range(len(data_list)):
                    if data_list[i] == 'HTTP':
                        destination = data_list[i-1]
                    else:
                        continue
                #avoid favicon
                if destination == 'favicon.ico':
                    skt.close()
                    listenlist.remove(skt)
                    continue
                result[destination] = 'null'
                s[skt] = destination
                #create data_send,This data means find route
                data_send=''
                #The format is: 1; destination; route
                data_send = str(1) + ';' + destination + ';' + station_name + ',' + str(udp_port)
                for i in neighbour_port:
                    print('send to:'+ str(i))
                    print('send message is：'+ data_send)
                    udpsocket.sendto(data_send.encode('utf-8'),(localhost,i))
                #skt remove from listen append write
                writelist.append(skt)
                listenlist.remove(skt)
        for skt in write:
            destination = s.get(skt)
            if result.get(destination)=='null':
                #not prepare for write
                continue
            elif result.get(destination)!='null':
                #http response
                output = result.get(destination)
                responseHeaderLines = "HTTP/1.1 200 OK\r\n"
                responseHeaderLines += "\r\n"
                responseBody = output
                response = responseHeaderLines + responseBody
                skt.send(response.encode('utf-8'))
                writelist.remove(skt)
                skt.close()
                
    udpsocket.close()
    tcpsocket.close()
if __name__ == '__main__':
    main()
                

                    


                

